### DISCUSSION TITLE
Different data type returned depending on serialization format

### DISCUSSION THREAD

**Discussion by ddeboer** [comunica/comunica#871] [Q&A]
When querying JSON-LD documents, any Schema.org date field is coerced to a `schema:Date` automatically. For example a `schema:dateCreated` value becomes a `schema:Date`:

```
$ npx comunica-sparql-file https://lab.coret.org/register/kb-dates.jsonld "CONSTRUCT WHERE { ?s <http://schema.org/dateCreated> ?o }"
<http://data.bibliotheken.nl/id/dataset/rise-alba> <http://schema.org/dateCreated> "2021-05-27"^^<http://schema.org/Date>.
```

When querying other serialization formats, however, this is not the case:

```
$ npx comunica-sparql-file https://lab.coret.org/register/kb-dates.ttl "CONSTRUCT WHERE { ?s <https://schema.org/dateCreated> ?o }"
<http://data.bibliotheken.nl/id/dataset/rise-alba> <https://schema.org/dateCreated> "2021-05-27".
```

I’ve seen that for JSON-LD documents, Comunica fetches the Schema.org vocabulary and apparently uses that to coerce values. Should this be the case for other serializations, too?

It may also be interesting to note that both source documents use `https://schema.org` (not `http://schema.org`) but the JSON-LD is returned as HTTP; I assume this is because Comunica redirects to `http://schema.org`.

**Comment by rubensworks** [comment]
Comunica doesn't do any kind of coercion/reasoning (at the moment), and it also does not do any special conversions for schema.org.
So this is likely a problem in the data source.

This seems to be confimed when looking at the raw parsing of the two files:

* JSON-LD: https://rdf-play.rubensworks.net/#url=https%3A%2F%2Flab.coret.org%2Fregister%2Fkb-dates.jsonld&proxy=https%3A%2F%2Fproxy.linkeddatafragments.org%2F
* Turtle: https://rdf-play.rubensworks.net/#url=https%3A%2F%2Flab.coret.org%2Fregister%2Fkb-dates.ttl&proxy=https%3A%2F%2Fproxy.linkeddatafragments.org%2F

FYI, the JSON-LD context of schema.org (https redirects to http) uses `http://` as prefix for its JSON-LD terms. Therefore, I would recommend to update your Turtle file to also use the http-based prefixes.

**Reply by ddeboer** [reply]
Hey @rubensworks, thanks for investigating!

How do you then explain the difference between:

- the JSON-LD document, which returns a date type:
  -   `<http://schema.org/dateCreated> "2021-05-27"^^<http://schema.org/Date>` for `schema:dateCreated "2021-05-27"`;
- the Turtle document, which returns a string type:
  -  `<https://schema.org/dateCreated> "2021-05-27"` for `"dateModified": "2021-05-27T09:56:21.370767"`?

The only reason I can come up with, is that Comunica takes the Schema.org context into account for JSON-LD documents but not for Turtle files.

> FYI, the JSON-LD context of schema.org (https redirects to http) uses http:// as prefix for its JSON-LD terms. Therefore, I would recommend to update your Turtle file to also use the http-based prefixes.

True! Unfortunately, we have to accept both `http://` and `https://schema.org` because the source data is not controlled by us.

**Comment by rubensworks** [comment]
> How do you then explain the difference between:

The Schema.org JSON-LD context probably contains a `@type` entry for `dateModified`, which sets the literal's datatype. The Turtle document (which has no relation to the Schema.org JSON-LD context), contains no datatype on the date literals.

> True! Unfortunately, we have to accept both http:// and https://schema.org because the source data is not controlled by us.

Then one of the following will probably be needed:

1. Create a proxy source over these sources that normalize the Schema.org URLs.
2. Normalize URLs in the SPARQL query (will probably have a big impact on performance).
3. Add (a very lightweight) reasoning component to Comunica (related to https://github.com/comunica/comunica/issues/533)

**Reply by ddeboer** [reply]
For reference, this is how we did this in our application:

- https://github.com/netwerk-digitaal-erfgoed/dataset-register/blob/main/src/transform.ts#L11-L52
- https://github.com/netwerk-digitaal-erfgoed/dataset-register/pull/309


Most probably a suboptimal solution (particularly the UNION SPARQL query).
